File: swift/FantasyBattle/Player.swift
struct Player: Target {
    let inventory: Inventory, stats: Stats
    init(inventory: Inventory, stats: Stats) {
        self.inventory = inventory
        self.stats = stats
    }
    
    func calculateDamage(other: Target) -> Damage {
        let baseDamage = getBaseDamage()
        let damageModifier = getDamageModifier()
        let totalDamage = baseDamage * damageModifier
        let soak = getSoak(other: other, totalDamage: totalDamage)
        return Damage(amount: max(0, totalDamage - soak))
    }
    
    func getBaseDamage() -> Double {
        let equipment = inventory.equipment
        let leftHand = equipment.leftHand
        let rightHand = equipment.rightHand
        let head = equipment.head
        let feet = equipment.feet
        let chest = equipment.chest
        return leftHand.getBaseDamage() +
                rightHand.getBaseDamage() +
                head.getBaseDamage() +
                feet.getBaseDamage() +
                chest.getBaseDamage()
    }
    
    func getDamageModifier() -> Double {
        let equipment = inventory.equipment
        let leftHand = equipment.leftHand
        let rightHand = equipment.rightHand
        let head = equipment.head
        let feet = equipment.feet
        let chest = equipment.chest
        let strengthModifier = stats.strength * 0.1
        return strengthModifier +
                leftHand.getDamageModifier() +
                rightHand.getDamageModifier() +
                head.getDamageModifier() +
                feet.getDamageModifier() +
                chest.getDamageModifier()
    }
    
    func getSoak(other: Target, totalDamage: Double) -> Double {
        var soak = 0.0
        if let player = other as? Player {
            // TODO: Not implemented yet
            //  Add friendly fire
            soak = totalDamage
        } else if let simpleEnemy = other as? SimpleEnemy {
            soak =
                simpleEnemy.getArmor().getDamageSoak() *
                simpleEnemy.getBuffs().map {$0.soakModifier()}.reduce(1, +)
        }
        return soak
    }
    
}

File: csharp/FantasyBattle/Player.cs
using System;
using System.Linq;

namespace FantasyBattle
{
    public class Player : Target
    {
        public Inventory Inventory { get; }
        public Stats Stats { get; }

        public Player(Inventory inventory, Stats stats)
        {
            Inventory = inventory;
            Stats = stats;
        }

        public Damage CalculateDamage(Target other)
        {
            int baseDamage = CalculateBaseDamage();
            float damageModifier = CalculateDamageModifier();
            int totalDamage = (int)Math.Round(baseDamage * damageModifier, 0);
            int soak = GetSoak(other, totalDamage);
            return new Damage(Math.Max(0, totalDamage - soak));
        }

        private int CalculateBaseDamage()
        {
            Equipment equipment = Inventory.Equipment;
            Item leftHand = equipment.LeftHand;
            Item rightHand = equipment.RightHand;
            Item head = equipment.Head;
            Item feet = equipment.Feet;
            Item chest = equipment.Chest;
            return leftHand.BaseDamage +
                   rightHand.BaseDamage +
                   head.BaseDamage +
                   feet.BaseDamage +
                   chest.BaseDamage;
        }

        private float CalculateDamageModifier()
        {
            Equipment equipment = Inventory.Equipment;
            Item leftHand = equipment.LeftHand;
            Item rightHand = equipment.RightHand;
            Item head = equipment.Head;
            Item feet = equipment.Feet;
            Item chest = equipment.Chest;
            float strengthModifier = Stats.Strength * 0.1f;
            return strengthModifier +
                   leftHand.DamageModifier +
                   rightHand.DamageModifier +
                   head.DamageModifier +
                   feet.DamageModifier +
                   chest.DamageModifier;
        }

        private int GetSoak(Target other, int totalDamage)
        {
            int soak = 0;
            if (other is Player)
            {
                // TODO: Not implemented yet
                //  Add friendly fire
                soak = totalDamage;
            }
            else if (other is SimpleEnemy simpleEnemy)
            {
                soak = (int)Math.Round(
                    simpleEnemy.Armor.DamageSoak *
                    (
                        simpleEnemy.Buffs.Select(x => x.SoakModifier).Sum() + 1
                    ), 0
                );
            }

            return soak;
        }
    }
}